File: examples/counter/web-yew/src/http.rs
use anyhow::Result;
use gloo_net::http;

use shared::http::protocol::{HttpRequest, HttpResponse};

pub async fn request(
    HttpRequest {
        method,
        url,
        headers,
        ..
    }: &HttpRequest,
) -> Result<HttpResponse> {
    let mut request = match method.as_str() {
        "GET" => http::Request::get(url),
        "POST" => http::Request::post(url),
        _ => panic!("not yet handling this method"),
    };

    for header in headers {
        request = request.header(&header.name, &header.value);
    }

    let response = request.send().await?;
    let body = response.binary().await?;

    Ok(HttpResponse::status(response.status()).body(body).build())
}

File: examples/counter-next/Android/app/src/main/java/com/crux/example/counter/http.kt
package com.crux.example.counter

import com.crux.example.counter.shared.HttpHeader
import com.crux.example.counter.shared.HttpRequest
import com.crux.example.counter.shared.HttpResponse
import com.novi.serde.Bytes
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.headers
import io.ktor.client.request.request
import io.ktor.http.HttpMethod
import io.ktor.util.flattenEntries

suspend fun requestHttp(
        client: HttpClient,
        request: HttpRequest,
): HttpResponse {
    val response =
            client.request(request.url) {
                this.method = HttpMethod(request.method)
                this.headers {
                    for (header in request.headers) {
                        append(header.name, header.value)
                    }
                }
            }
    val bytes = Bytes.valueOf(response.body())
    val headers = response.headers.flattenEntries().map { HttpHeader(it.first, it.second) }
    return HttpResponse(response.status.value.toShort(), headers, bytes)
}
